### DISCUSSION TITLE
Mapping EMS/ESP32 sensors for SAT

### DISCUSSION THREAD

**Discussion by sergeantd83** [emsesp/EMS-ESP32#2257] [Ideas]
Following this issue https://github.com/emsesp/EMS-ESP32/issues/2103, we are doing an initially effort to map the EMS/ESP sensors for SAT usage.

**Part I ( Mapping EMS/ESP32 sensors for SAT )**

| SAT | EMS Sensor | UOM | Read / Write | Description |Mandatory Values |
| --- | :---: |:---:|:---:|:---:|:---:|
| DATA_FLAME_ACTIVE | "burngas" | Binary |R| Whether the flame is ON or OFF|✔️|
| DATA_DHW_SETPOINT | "dhw/seltemp" | °C |R / W| The DHW setpoint |✔️|
| DATA_CONTROL_SETPOINT | "selflowtemp" | °C |R / W| The Boiler flow water temperature setpoint |✔️|
| DATA_REL_MOD_LEVEL | "curburnpow" | % |R| The current boiler power |✔️|
| DATA_BOILER_TEMPERATURE | "curflowtemp" | °C |R| The boiler flow water temperature |✔️|
| DATA_RETURN_TEMPERATURE | "rettemp" | °C |R| The return water temperature |✔️|
| DATA_DHW_ENABLE | "tapwateractive" | Binary |R| When the DHW is in use |✔️|
| DATA_CENTRAL_HEATING | "heatingactive" | Binary |R| When the thermostat asks for heating |✔️|
| DATA_SLAVE_MEMBERID | - | Integer| R| The unique number of the Boiler Manufacturer |❌|
| DATA_BOILER_CAPACITY | "nompower" | kW |R| The Rated maximum power in kW of the boiler |❌|
| DATA_REL_MIN_MOD_LEVEL | "burnminnpower" | % |R| |❌|
| DATA_MAX_REL_MOD_LEVEL_SETTING | "burnmaxpower" | % |R / W| The maximum allowable power of the boiler |✔️|
| DATA_DHW_SETPOINT_MINIMUM | - | °C | R| The min value for the DHW setpoint |❌ ( But nice to have ) |
| DATA_DHW_SETPOINT_MAXIMUM |  -  | °C |R| The max value for the DHW setpoint |❌ (But nice to have) |

**Part II ( Boiler control )**

| OTGW Command Payload | EMS/ESP32 Command Payload |  Description |
| :---: | :---: |:---:|
| ```await self._send_command(f"CH={1 if state == DeviceState.ON else 0}")``` | - | The CH command starts / stops the heating |
| ```await self._send_command(f"CS={value}")``` | ```await self._send_command({"cmd":"selflowtemp","value":0 if value == 10 else value})```  | The CS command manipulates the boiler flow water setpoint |
| ```await self._send_command(f"MM={value}")```| ```await self._send_command({"cmd":"burnmaxpower","value":value})```| The MM command manipulates the maximum modulation of the boiler |
| ```await self._send_command(f"SW={value}")``` | ```await self._send_command({"cmd":"dhw/seltemp","value":value})```| The SW command changes the DHW setpoint |
| ```await self._send_command(f"SH={value}")``` |  ```await self._send_command({"cmd":heatingtemp","value":value})``` | The SH command manipulates the max allowable boiler flow water temperature |

**Part III ( Command Topic )**

| OTGW Command Topic | EMS/ESP32 Command Topic |
| :---: | :---: |
| ```await mqtt.async_publish(self.hass, f"{self._topic}/set/{self._node_id}/command", payload)``` | ```await mqtt.async_publish(self.hass, f"ems-esp/<base>/boiler", payload)``` |

**Part IV ( Publish Topic )**

| OTGW Publish Topic | EMS/ESP32 Publish Topic |
| :---: | :---: |
| ```OTGW/value/{self._node_id}``` | ```ems-esp/boiler_data``` |

**Comment by proddy** [comment]
Could you add the unit of measure (uom) and a description of the SAT name? That will help us locate the EMS variant. Some of the names are very ambiguous, like "DATA_CENTRAL_HEATING" ? Also `tapwateractive` and `heatingactive` are not native EMS values as they are calculated by the EMS-ESP software based on boiler states, heating type and current flow temperatures, and not 100% accurate. So we may need to find alternatives if this is essential for SAT.

**Reply by proddy** [reply]
@MichaelDvP will need your wisdom here too as it progresses

**Comment by MichaelDvP** [comment]
Edit the table:
DATA_FLAME_ACTIVE 	"burngas" 	Binary 	R 	Whether the flame is ON or OFF 	✔️
DATA_DHW_SETPOINT 	"dhw/seltemp"	°C 	R / W 	The DHW setpoint 	✔️
DATA_CONTROL_SETPOINT 	"selflowtemp" 	°C 	R / W 	The Boiler flow water temperature setpoint 	✔️
DATA_REL_MOD_LEVEL 	"curburnpow" 	% 	R 	The current boiler power 	✔️
DATA_BOILER_TEMPERATURE 	"curflowtemp" 	°C 	R 	The boiler flow water temperature 	✔️
DATA_RETURN_TEMPERATURE 	"rettemp" 	°C 	R 	The return water temperature 	✔️
DATA_DHW_ENABLE 	"tapwateractive"	Binary 	R 	When the DHW is in use 	✔️
DATA_CENTRAL_HEATING 	"heatingactive"	Binary 	R 	When the thermostat asks for heating 	✔️
DATA_SLAVE_MEMBERID 		Integer 	R 	The unique number of the Boiler Manufacturer 	❌
DATA_BOILER_CAPACITY 	"nompower"	kW 	R/W 	The Rated maximum power in kW of the boiler 	❌
DATA_REL_MIN_MOD_LEVEL 	"burnminnpower" 	% 	R / W 		❌
DATA_MAX_REL_MOD_LEVEL_SETTING 	"burnmaxpower" 	% 	R / W 	The maximum allowable power of the boiler 	✔️
DATA_DHW_SETPOINT_MINIMUM 	30	°C 	R 	The min value for the DHW setpoint 	❌ ( But nice to have )
DATA_DHW_SETPOINT_MAXIMUM 	60	°C 	R 	The max value for the DHW setpoint 	❌ (But nice to hav

For DATA_DHW_SETPOINT we have also "dhw/settemp", readonly. This could be lower then dhw/seltemp if solar is active and wants to prevent dhw charging from boiler. But for controlling dhw/seltemp is right to read and write.
"tapwateractive" and "heatingactive" are calculated by ems-esp, but should work for all boilers. 
For DHW there are flow systems, showing "tapwateractive" if there is flow and burner is on, for buffer systems "tapwateractive" shows charging of the dhw buffer (with burner on).
For DATA_BOILER_CAPACITY we have "nompower", but not published by all boilers. Also sometimes it is the power for dhw charging sometimes for heating. We have it editable.
DATA_DHW_SETPOINT_MAXIMUM is "dhw/maxtemp" for some boilers, but often not published.
burnminpower, burnmaxpower are only writeable for modulated boilers, otherwise fixed. I'm not sure how 2-stage burners handle burngas, maybe they switch to burngas2, maybe both are set.

**Comment by sergeantd83** [comment]
@MichaelDvP, @proddy thanks! 
I added a Part II table in the original post about how we control the boiler. Is it helpful to match the commnads you send to the EMS gateway?

**Reply by MichaelDvP** [reply]
> The CH command starts / stops the heating

There is no ems command for this. We have an artificial setting `forceheatingoff`, which sets `selflowtemp` to zero and repeats this every minute. 
The ems-boilers have a setting `heatingtemp`, for most ems-boilers set on the controller with mechanical Poti, on ems+ boilers also via bus changeable. The boiler defaults to this temperature if there is not periodical `selflowtemp` command, timeout is ~ 1,5 min. Also `selflowtemp` is limited to be maximal `heatingtemp`. 
> The CS command manipulates the boiler flow water setpoint

`selflowtemp`
> The MM command manipulates the maximum modulation of the boiler

`burnmaxpower`
> The SW command changes the DHW setpoint

`dhw/seltemp`
> The SH command manipulates the max allowable boiler flow water temperature

`heatingtemp` (sometimes fixed on the boiler control).

The commands are all on device boiler. Using mqtt the topic is `ems-esp/<base>/boiler/selflowtemp` with payload `{"value":xx}`. (ems-esp can be renamed, base is optional),
Using the api it is POST: `http://ems-esp.local/api/boiler/selflowtemp` with body `{"value":xx}`.

**Reply by sergeantd83** [reply]
If the `selflowtemp` is set to 10, does the boiler stop the pump?

Because when SAT stops the heating,  sends the value 10 every 30 seconds.

Also SAT sends every 30 seconds the calculated Control Setpoint ( selflowtemp ) so there is no problem for the boiler to retain the value.

**Reply by MichaelDvP** [reply]
Don't know if the boiler stops the pump with setpoint 10. My pump is controlled by a mixer and thermostat, so can't test without reconfiguring the complete system. We will need someone with only boiler to test this out.

**Reply by sergeantd83** [reply]
So when the thermostat stops the heating just sends the value 0?

**Reply by MichaelDvP** [reply]
Yes. 
My RC35 thermostat stops heating (test by setting  setting summermode) by sending selflowtemp 0 and burnmaxpower 0, and a third 0 for pump modulation (it's in the same ems telegram, but on my system with mixer it is always 0). Also the ems-esp command forceHeatingOff sets these 3 parameters at once. Actually we don't have a command to set the pump modulation by this telegram (0x1A).
Maybe better use the forceHeatingOff=true to stop heating and forceHeatingOff=false to restart. This is confirmed working for some people controlling the boiler direct.

But we can also add some commands/entities for SAT if this helps.

**Reply by sergeantd83** [reply]
Thanks for the detailed informations. I wonder if you could add a new topic for commads, something like:
 `ems-esp/<base>/boiler/command` and payloads:
 
- `selflowtemp:xx`
- `burnmaxpower:xx`
- `dhwseltemp:xx`
- `heatingtemp:xx`

As a result SAT will send the below payloads:

- `selflowtemp`: `await self._send_command(f"selflowtemp:{value}")`
- `burnmaxpower`:`await self._send_command(f"burnmaxpower:{value}")`
- `dhw/seltemp`:`await self._send_command(f"dhwseltemp:{value}")`
- `heatingtemp`: ``await self._send_command(f"heatingtemp:{value}")``

I think it is fine for SAT to manipulate only these values, there is no need to use the `forceHeatingOff` command.
 
Using this topic/payload stracture, EMS/ESP32 commands align with the SAT OTGW commands.

This is the topic we send the commands to the OTGW:
<https://github.com/Alexwijn/SAT/blob/f9a12f8df9d7f548378329acbc8682f7620238be/custom_components/sat/mqtt/__init__.py#L243>

**Reply by proddy** [reply]
We have those commands already, see the documentation https://docs.emsesp.org/Commands/#mqtt

**Reply by MichaelDvP** [reply]
> I wonder if you could add a new topic for commads, something like:
ems-esp/<base>/boiler/command and payloads:
    selflowtemp:xx

ems.esp accept different formats, you can also use topic `ems-esp/<base>/boiler` and payload `{"cmd":"selflowtemp","value":xx}`
When using full topic `ems-esp/<base>/boiler/selflowtemp` the payload can be a single number or a json `{"value":xx}`

If we add this `command`-command, is the payload a json pair `{"selflowtemp":xx}` or just a string `selflowtemp:xx`?
Is it garanteed that payload contains a single pair? Or do we need to loop the pairs and generate multiple commands?

BTW: i rechecked the commands. `burnMaxPow` is an eeprom setting in the boiler as general upper limit, when sending periodical the actual ram value is `selBurnPow`

**Reply by proddy** [reply]
I prefer not to add more code to ems-esp and to adjust the SAT python code which is a Home Assistant add-on, or even better do a specific fork that doesn't use OpenTherm at all. The only logic we need is the self-learning heating curve.

**Reply by sergeantd83** [reply]
> I prefer not to add more code to ems-esp and to adjust the SAT python code which is a Home Assistant add-on

Ok that was the original idea. I just asked because @MichaelDvP kindly offered.

> or even better do a specific fork that doesn't use OpenTherm at all. 

We don't care to create an "EMS-SAT"fork. Because SAT is a universal controller and an on-going development. It not our intention to maintain 2 repositories.

> The only logic we need is the self-learning heating curve.

Probably you misunderstood SAT. The heating curve calculation is only one of the ongoing calculations and actually the easiest one. The hard part is the boiler control in the most efficient way while respecting the heating system limitations.

**Reply by Delta1977** [reply]
That would be a great addition to Ems-esp. I hope the implementation is going well—I can’t wait to try it out. I’d also be happy to be a beta tester!

**Reply by MichaelDvP** [reply]
> I just asked because @MichaelDvP kindly offered.

Yes, ems-esp has a lot of code and special entities to adapt to the dump HA climate widget. 
So i assumed that we can add a few lines to allow the different command format for the more powerfull SAT. 
Sorry if i have promised too much.

**Reply by sergeantd83** [reply]
No worries, our job is to create the EMS coordinator without mistakes 🙂. Yesterday we worked with @Alexwijn and I think we finished the mapping. Can you please review the tables in the original post if they are correct? It will be a huge help since we can’t test anything.

Thanks again for your help!

**Comment by proddy** [comment]
@MichaelDvP we need your expertise here. On Discord there was a discussion on how the SAT would send the commands similar to the OpenTherm GW which uses heating on/off and setpoint values. I explained in EMS the boiler will only except a new setpoint temperature value if it's lower than the actual current heating temperature, which is why we have the `forceheatingoff` command to send 0's every minute. So for SAT to work with EMS-ESP this Setting needs to be enabled and SAT only needs to send MQTT topics `ems-esp/boiler/selflowtemp`.

Is this a correct understanding?

@sergeantd83 mention also

> OTGW intercepts the Thermostat messages and prioritizes the published commands. When there is not a new command for 60 seconds then falls back to thermostat operation. With this solution we can have a back up OT thermostat while SAT controls the heating when network and HA are up and running. 
CS=temperature
Control Setpoint — Manipulate the control setpoint being sent to the boiler. Set to 0 to pass along the value specified by the thermostat. To stop the boiler heating the house, set the control setpoint to some low value and clear the CH enable bit using the CH command. A CS command with a value of 8 or higher must be repeated at least every minute as long as adjustment is needed. This is a vigilance check to prevent runaway heating in case the controlling program loses its connection, or crashes.
Warning: manipulating these values may severely impact the control algorithm of the thermostat, which may cause it to start heating much too early or too aggressively when it is actually in control.
Examples: CS=45.8, CS=0

**Reply by MichaelDvP** [reply]
`forceheatingoff` overrides any `selflowtemp` setting, so has to be `off` to control the boiler.
You can't use SAT or direct boiler control if a ems(+)-thermostat is connected. The thermostat will overwrite the any selflowtemp setting. A fallback if HA/SAT is disconnected/down is the boiler setting `heatingtemp` that become active when there is no cyclic selflowtemp command .On many boilers the `heatingtemp` is set by a dial and can't be changed via ems-bus.
I don't know if a low selflowtemp value like 8°C is possible. Someone without thermostat have to check.

**Reply by sergeantd83** [reply]
> forceheatingoff overrides any selflowtemp setting, so has to be off to control the boiler.

Thanks for the confirmation. It is included in the commands [list](https://github.com/Alexwijn/SAT/blob/fbb87b71f754d2a3655c517ec591c527470e57d7/custom_components/sat/mqtt/ems.py#L121).

> I don't know if a low selflowtemp value like 8°C is possible. Someone without thermostat have to check.

We use [this](https://github.com/Alexwijn/SAT/blob/fbb87b71f754d2a3655c517ec591c527470e57d7/custom_components/sat/mqtt/ems.py#L107) logic for the `selflowtemp` command.

> You can't use SAT or direct boiler control if a ems(+)-thermostat is connected. The thermostat will overwrite the any selflowtemp setting.

Who knows, you may found a solution for EMS/ESP32 firmware to intercept the thermostat messages to the boiler. Then SAT can work on top of an existing EMS thermostat 😄.

Thank you very much for the information. @MichaelDvP and @proddy, without your help, the EMS coordinator would have been impossible.

**Reply by csutcliff** [reply]
> I don't know if a low selflowtemp value like 8°C is possible. Someone without thermostat have to check.

My HT3 system will accept down to 11°C. Writing 10°C or below it goes to 0°C instead.

@sergeantd83 is DATA_RETURN_TEMPERATURE completely mandatory? Some boilers (mine included) do not have the "rettemp" available.

If it is absolutely required, could it be provided by a temp probe stuck to the return pipe? These sensors are natively supported in EMS-ESP so it would just need some setting to choose the correct sensor.

**Reply by sergeantd83** [reply]
It is not necessary to have this value reported. SAT can handle the calculations without  this value. We are also working on a new feature to dynamically change the selflowtemp based on the boiler flow water temperature, a feature that will be quite beneficial for boilers that don’t report the return water temperature and TRVs are installed.

**Reply by MichaelDvP** [reply]
@sergeantd83  You have set return temp as mandatory, but some bosch boilers does not have a sensor connected. In this cases `rettemp` is missing in publish payload. If SAT needs it to work, could there be a fallback to a DS1820 temperaturesensor? We could advice users to connect a DS1820 and name it `rettemp`, then it is published with topic `temperaturesensor_data`, payload as normally in boiler_data: `{"rettemp":xx}` 

Also possible: we can use custom entities in RAM (topic `custom_data`, payload `{"rettemp":xx}` with fixed names and users have to create on-chang-schedules to fill them with values.

**Reply by sergeantd83** [reply]
That would be a great addition! Then SAT will be able to monitor the return temperature [here](https://github.com/Alexwijn/SAT/blob/fbb87b71f754d2a3655c517ec591c527470e57d7/custom_components/sat/mqtt/ems.py#L19). This is the true power of open source and cooperation.

**Comment by proddy** [comment]
@MichaelDvP do you know if we can get the minimum power from the boiler that it needs to modulate? Like we do with `nompower` for the maximum power. We get the min and max from x04/UBAFactory and use this to set the min/max range for burnMaxPower but I think that is a % and not a kW, although we can convert it.

**Reply by sergeantd83** [reply]
We need this value in % not kW. So the min value from x04/UBAFactory is just fine.

**Reply by MichaelDvP** [reply]
telegram 0x04 is not implemened in all boilers. 
The actual boiler setting is `burnMinPower` from telegram 0x16(ems) or 0xE6(ems+).
It's also in percent.

**Reply by csutcliff** [reply]
This also needs to be able to be manually set. My boiler reports 0% as the minimum but in actuality it can only go down to 30%

**Reply by proddy** [reply]
Same with my Nefit boiler. `burnMinPower` value is 0, unless manually set, and the actual min power is from the UBAFactory telegram which shows a value of 23 (%) which also matches the min value in the boiler's manual.

**Reply by MichaelDvP** [reply]
Then i think we have to add an entity for factoryMinPower from 0x04. 
My burner is single stage oil and the boiler could be equipt with different burner types, so there is no nompower.
In 0x04 i have min=100%, max=100%, nompower=0 and indeed burnMinPower from 0x16 is 0. 
I don't know what `burnMinPower` is used for, i thought it was the min setting, but i was wrong.

**Comment by jlk3281** [comment]
@sergeantd83 Would there be a chance to reconsider the topic based on https://github.com/emsesp/EMS-ESP32/discussions/2641#discussioncomment-14624765?
